### PR TITLE
Restarts shouldn't need to fetch lastSuccessfulMeta again

### PR DESCRIPTION
**Pull request by scr-oath**
## Context

Support for rolling back a deployment to production is via finding the last good deployment and restarting the job.  If such a job fetched lastSuccessfulMeta from an external job, it would fetch the newest state - not the state fetched when it first run.

In order to support reproducible behavior, store the external metadata in a key of the main meta so that it is then stored with the event.

Follow up to #24, this stores external metadata in a key `sd.$pipelineId.$jobName` corresponding to metadata for `sd@$pipelineId:$jobName`.

### Note: increased scope while implementing:
The logic in getMeta was getting fairly large, as was its positional parameter list. To make it more discrete, made a `MetaSpec` object with smaller pointer receivers (aka methods) that perform smaller tasks and call each other.

Because this overhaul required touching many, if not all, of the tests, I also overhauled the tests to use a `suite.Suite` and to share common code with a testCases style of sub-tests.  This let me also split the mocked http responses to a file for sharing between the multiple unit tests, and slurping into members of the suites at `SetupSuite` invocation.

## Objective

To support reproducibility of restarting a job which grabs the _SNAPSHOT_ of metadata from an external job when it first runs (or is triggered by) that external job, and does _NOT_ fetch any metadata on subsequent runs (such as restarting for the purpose of deployment of an older version to production).

## References

#24

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: meta.go [hunk 1]
```diff
@@ -59,9 +61,23 @@ func getMeta(key string, metaSpace string, metaFile string, output io.Writer, js
...
+		if err == nil && sb.String() != "null" {
+			logrus.Debugf("Found data in external meta key %s", externalMetaKey)
+			data = []byte(sb.String())
+		} else {
+			logrus.Debugf("Fetching metadata from %s", jobDescription.External())
+			data, err = lastSuccessfulMetaRequest.FetchLastSuccessfulMeta(jobDescription)
+			if err != nil {
+				return err
+			}
+			err = setMeta(externalMetaKey, string(data), metaSpace, kDefaultMetaFile, true)
```
Comment (scr-oath) [inline meta.go]: Do we also need to set this when read from file?  Will screwdriver launcher take care of setting this metadata for meta?

Comment (scr-oath) [reply meta.go]: Done with a refactor to make it slightly easier to write.

File: internal/fetch/jobDescription_test.go [hunk 2]
```diff
@@ -1,13 +1,19 @@
 package fetch
 
 import (
+	"github.com/stretchr/testify/suite"
```
Comment (dmitris) [inline internal/fetch/jobDescription_test.go]: could you please put it below stdlib imports (separated with an empty line), `goimports`-compatible?
```
import (
	"testing"

	"github.com/stretchr/testify/suite"
)
```

File: internal/fetch/jobDescription_test.go [hunk 3]
```diff
@@ -33,14 +39,54 @@ func Test_parseJobDescription(t *testing.T) {
...
+			name: "sd@123:fooBar",
+			jobDescription: JobDescription{
+				PipelineID: 123,
+				JobName:    "fooBar",
+			},
+		},
+	} {
+		s.Run(tc.name, func() {
+			got := tc.jobDescription.External()
+			s.Assert().Equal(tc.name, got)
```
Comment (dmitris) [inline internal/fetch/jobDescription_test.go]: (nit) the recommended order is `got, then want` - both in comparisons as well as asserts and error messages

File: meta.go [hunk 4]
```diff
@@ -29,51 +35,123 @@ var (
...
 
-// getMeta prints meta value from file based on key
-func getMeta(key string, metaSpace string, metaFile string, output io.Writer, jsonValue bool, lastSuccessfulMetaRequest *fetch.LastSuccessfulMetaRequest) error {
-	metaFilePath := metaSpace + "/" + metaFile + ".json"
+// MetaSpec encapsulates the parameters usually from CLI so they are more readable and shareable than positional params.
+type MetaSpec struct {
+	MetaSpace                    string
+	SkipFetchNonexistentExternal bool
+	MetaFile                     string
+	JsonValue                    bool
```
Comment (dmitris) [inline meta.go]: consider adding a comment regarding what this 'JsonValue' is or should be used for or what it contains.

File: meta.go [hunk 5]
```diff
@@ -335,15 +412,12 @@ func setMetaValueRecursive(key string, value string, previousMeta interface{}, j
 
 // setupDir makes directory and json file for meta
 func setupDir(metaSpace string, metaFile string) error {
```
Comment (dmitris) [inline meta.go]: `staticcheck` [1] says: `meta.go:414:6: func setupDir is unused (U1000)`.  I think you have `MetaSpec.SetupDir` now - can this be deleted?

```
$ go doc github.com/screwdriver-cd/meta-cli.MetaSpec.SetupDir
func (m *MetaSpec) SetupDir() ([]byte, error)
    SetupDir creates the metaspace directory and writes a file with empty
    object.
```

[1] http://staticcheck.io/docs/

File: meta_test.go [hunk 6]
```diff
@@ -1,216 +1,223 @@
...
+	// Test get
+	Require.NoError(s.CopyMockFile(externalFile))
 
-	stdout = new(bytes.Buffer)
-	getMeta("ary[2].ccc.ddd[1]", mockDir, testFile, stdout, false, nil)
-	expected = []byte("2")
-	if bytes.Compare(expected, stdout.Bytes()) != 0 {
-		t.Fatalf("not matched. expected '%v', actual '%v'", string(expected), string(stdout.Bytes()))
-	}
+	got, err := s.MetaSpec.Get("str")
```
Comment (dmitris) [inline meta_test.go]: `staticcheck`: `meta_test.go:96:7: this value of err is never used (SA4006)` - should the test check if `err` is not nil?

File: meta_test.go [hunk 7]
```diff
@@ -844,34 +629,141 @@ func TestSymmetry_json_object(t *testing.T) {
...
+			Require.NoError(os.Remove(testFilePath))
 
 			// Set and get the jsonValue to/from writable file with jsonValue true
-			require.NoError(t, setMeta(tc.key, jsonValue, testDir, testFile, true))
-			stdout = new(bytes.Buffer)
-			require.NoError(t, getMeta(tc.key, testDir, testFile, stdout, true, nil))
-			newJsonValue := stdout.String()
-			assert.Equal(t, jsonValue, newJsonValue)
+			Require.NoError(s.MetaSpec.Set(tc.key, jsonValue))
+			newJsonValue, err := s.MetaSpec.Get(tc.key)
```
Comment (dmitris) [inline meta_test.go]: `statuccheck`: `meta_test.go:659:18: this value of err is never used (SA4006)` - add checking the error

File: meta.go [hunk 8]
```diff
@@ -21,6 +22,11 @@ import (
 	"gopkg.in/urfave/cli.v1"
 )
 
+const (
+	kDefaultMetaFile  = "meta"
```
Comment (dmitris) [inline meta.go]: what does the leading "k" stand for?  `golint` says:
`meta.go:26:2: don't use leading k in Go names; const kDefaultMetaFile should be defaultMetaFile`

same with `kDefaultMetaSpace` on the next line.

File: meta.go (hunk 4 above)
Comment (dmitris) [reply meta.go]: also `golint` recommends: `meta.go:46:2: struct field JsonValue should be JSONValue`

File: meta.go [hunk 9]
```diff
@@ -82,27 +160,72 @@ func getMeta(key string, metaSpace string, metaFile string, output io.Writer, js
...
+		return fmt.Sprintf("%v", string(resultJSON)), nil
 	case nil:
-		_, err = fprintf(output, "null")
+		return "null", nil
 	default:
-		if jsonValue {
+		if m.JsonValue {
 			resultJSON, _ := json.Marshal(result)
-			_, err = fprintf(output, "%v", string(resultJSON))
+			return fmt.Sprintf("%v", string(resultJSON)), nil
```
Comment (dmitris) [inline meta.go]: `golint` regarding the next line - I translate for clarity 😄
`meta.go:178:10: 'if' block ends with a return statement, so drop the 'else' keyword and outdent its block`

**Comment by dmitris** [comment]
golint findings on the repo:
[golint.txt](https://github.com/screwdriver-cd/meta-cli/files/3627208/golint.txt)

File: internal/fetch/lastSuccessfulMeta_test.go [hunk 10]
```diff
@@ -1,231 +1,21 @@
 package fetch
 
 import (
+	"github.com/stretchr/testify/suite"
```
Comment (dmitris) [inline internal/fetch/lastSuccessfulMeta_test.go]: please reorder after stdlib - goimports-compatible

File: internal/fetch/jobDescription_test.go [hunk 11]
```diff
@@ -33,14 +39,54 @@ func Test_parseJobDescription(t *testing.T) {
...
+	}{
+		{
+			name: "sd@123:fooBar",
+			jobDescription: JobDescription{
+				PipelineID: 123,
+				JobName:    "fooBar",
+			},
+		},
+	} {
+		s.Run(tc.name, func() {
```
Comment (dmitris) [inline internal/fetch/jobDescription_test.go]: I would prefer to separate the `tc` declaration / initialization and the `for` loop for better readability and clarity:
```
tests := []struct {
  // struct fields
} {
   {name: "foo", jobDescription: /* etc */ },
   // fill in the slice with test instances
} // done with tests "creation"! Deep breath... :)

for _, tt := range tests {
  // run tests
}

File: meta_test.go [hunk 12]
```diff
@@ -1,216 +1,223 @@
 package main
 
 import (
-	"bytes"
+	"github.com/screwdriver-cd/meta-cli/internal/fetch"
```
Comment (dmitris) [inline meta_test.go]: please reorder imports - here's what I'm getting after pressing "Save" in my editor (with `goimports` integrated):
```
import (
	"io"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"

	"github.com/screwdriver-cd/meta-cli/internal/fetch"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"
	"github.com/termie/go-shutil"
)
```

**Comment by dmitris** [comment]
@scr-oath "PR-to-PR" to fix imports ordering: https://github.com/scr-oath/meta-cli/pull/2

File: internal/fetch/jobDescription_test.go (hunk 3 above)
Comment (scr-oath) [reply internal/fetch/jobDescription_test.go]: I'm sorry to say, that's wrong - it's expected, actual, msgAndArgs...

https://godoc.org/github.com/stretchr/testify/assert#Assertions.Equal

File: internal/fetch/jobDescription_test.go (hunk 11 above)
Comment (scr-oath) [reply internal/fetch/jobDescription_test.go]: Done

File: meta.go (hunk 8 above)
Comment (scr-oath) [reply meta.go]: Sorry… hadn't reviewed style but _had_ seen that elsewhere so was tainted… will fix.

Comment (scr-oath) [reply meta.go]: Done

File: meta.go (hunk 9 above)
Comment (scr-oath) [reply meta.go]: Done

File: meta.go (hunk 5 above)
Comment (scr-oath) [reply meta.go]: Yep; done

File: meta_test.go (hunk 6 above)
Comment (scr-oath) [reply meta_test.go]: Done

File: meta_test.go (hunk 7 above)
Comment (scr-oath) [reply meta_test.go]: Done

File: internal/fetch/jobDescription_test.go (hunk 2 above)
Comment (scr-oath) [reply internal/fetch/jobDescription_test.go]: Done

File: meta.go (hunk 4 above)
Comment (scr-oath) [reply meta.go]: addressed all golint warnings

File: internal/fetch/lastSuccessfulMeta_test.go (hunk 10 above)
Comment (scr-oath) [reply internal/fetch/lastSuccessfulMeta_test.go]: Ran goimports

File: internal/fetch/jobDescription_test.go (hunk 3 above)
Comment (dmitris) [reply internal/fetch/jobDescription_test.go]: ok - never mind then.  Too bad the `assert` library deviates from the Go conventions as per https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures as well as the example in https://golang.org/pkg/testing/ - `t.Errorf("Abs(-1) = %d; want 1", got)`...

File: internal/fetch/lastSuccessfulMeta_test.go [hunk 13]
```diff
@@ -236,8 +26,28 @@ func (m *MockHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	m.Called(w, r)
 }
 
-func TestLastSuccessfulMetaRequest_LastSuccessfulMetaURL(t *testing.T) {
-	for _, tc := range []struct {
+type LastSuccessfulMetaSuite struct {
+	suite.Suite
+	JobsJson               string
```
Comment (dmitris) [inline internal/fetch/lastSuccessfulMeta_test.go]: could you change it to `JobsJSON` to stick to the Go's (opinionated) convention of abbreviations / "initialisms"?  https://github.com/golang/go/wiki/CodeReviewComments#initialisms  (can do in a follow-up)

File: internal/fetch/jobDescription.go [hunk 14]
```diff
@@ -31,3 +32,11 @@ func ParseJobDescription(defaultPipelineID int64, external string) (*JobDescript
 	ret.JobName = matches[2]
 	return ret, nil
 }
+
+func (jd *JobDescription) External() string {
```
[the remark this answers is not in the document]
Comment (scr-oath) [reply internal/fetch/jobDescription.go]: Done

File: internal/fetch/lastSuccessfulMeta_test.go [hunk 15]
```diff
@@ -2,230 +2,20 @@ package fetch
...
-      }
-    ],
-    "pipelineId": 1016709,
-    "state": "ENABLED",
-    "archived": false
-  }
-]
-`
-	metaJson = `{"foo","bar","arr":[1,2,3]}`
+	kMockHttpDir            = "../../mockHttp"
```
[the remark this answers is not in the document]
Comment (scr-oath) [reply internal/fetch/lastSuccessfulMeta_test.go]: Yep missed one. Done

File: internal/fetch/jobDescription.go [hunk 16]
```diff
@@ -31,3 +32,11 @@ func ParseJobDescription(defaultPipelineID int64, external string) (*JobDescript
...
+	return fmt.Sprintf("sd@%d:%s", jd.PipelineID, jd.JobName)
+}
+
+func (jd *JobDescription) MetaKey() string {
```
[the remark this answers is not in the document]
Comment (scr-oath) [reply internal/fetch/jobDescription.go]: Done

**Pull request merged**